### PR TITLE
Mag l1a dupe packets

### PR DESCRIPTION
**Pull request by maxinelasp**
# Change Summary
Remove duplicate packets from MAG if they have the same SHCOARSE, SRC_SQ_CTR, and APID. 

Fixes a bug found in L1a. Also adds an error throw if the data is non-monotonically increasing, so bugs like this are easier to find (as MAG is prone to them.)

Note that we don't expect to need this in prod, but most other instruments do it in the packets_to_dataset function. 

## Overview

Added a step which adds the MagL0 object to a set and confirms it is unique. This required the __eq__ and __hash__ functions on the dataclass.

### CODE REVIEW

File: imap_processing/mag/l0/mag_l0_data.py [hunk 1]
```diff
@@ -116,3 +119,46 @@ def __post_init__(self) -> None:
 
         self.PRI_VECSEC = 2**self.PRI_VECSEC
         self.SEC_VECSEC = 2**self.SEC_VECSEC
+
+    def __eq__(self, other: object) -> bool:
```
Comment (greglucas) [inline imap_processing/mag/l0/mag_l0_data.py:123]: Can't you set `eq=True` in the dataclass decorator to get this automatically for you?

File: imap_processing/tests/mag/test_mag_decom.py [hunk 2]
```diff
@@ -118,3 +122,28 @@ def test_mag_raw_cdf_generation(cdf_attrs):
 
     input_xarray = load_cdf(output)
     assert input_xarray.attrs.keys() == burst_data.attrs.keys()
+
+
+def test_comparison():
```
Comment (greglucas) [inline imap_processing/tests/mag/test_mag_decom.py:127]: Would it make sense to also add a test that loads a ccsds file twice so that you know you are actually pruning duplicate packets? I think this is a nice unit test, but might also be nice to add the larger integration to show it is doing what you want and being used.

I pushed an example up here https://github.com/IMAP-Science-Operations-Center/imap_processing/pull/1850

File: imap_processing/mag/l0/decom_mag.py [hunk 3]
```diff
@@ -43,8 +43,11 @@ def decom_packets(packet_file_path: str | Path) -> dict[str, list[MagL0]]:
 
     packet_definition = definitions.XtcePacketDefinition(xtce_document)
 
-    norm_data = []
-    burst_data = []
+    # We use a set for de-duplication, but we also keep a list to keep objects in order.
```
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: This seems like a lot of workaround by adding two objects to keep track of things throughout. Can you use a dict which is guaranteed to preserve order instead so you only have one object to keep track of?

```python
norm_dict = {}
norm_dict[mag_l0] = None

# if you want a list at the end
return list(norm_dict.keys())
```

File: imap_processing/mag/l0/mag_l0_data.py (hunk 1 above)
Comment (maxinelasp) [reply imap_processing/mag/l0/mag_l0_data.py:123]: Since I was only comparing SHCOARSE, APID, and SQ_CTR I thought this would be more reliable. I think your idea would probably also require freezing the CcsdsHeader class, which Lo uses as well.

File: imap_processing/tests/mag/test_mag_decom.py (hunk 2 above)
Comment (maxinelasp) [reply imap_processing/tests/mag/test_mag_decom.py:127]: Sure, I can do that

File: imap_processing/mag/l0/decom_mag.py (hunk 3 above)
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: Sure, that seems like a good solution

File: imap_processing/mag/l0/decom_mag.py [hunk 4]
```diff
@@ -43,8 +43,9 @@ def decom_packets(packet_file_path: str | Path) -> dict[str, list[MagL0]]:
...
```
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Need to update this comment to be a dict for de-duplication now.

**Review by greglucas — APPROVED** [review]
